File: 49-group-anagrams/49-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        
       
        // making a hashmap to easily add values to keys with the same letters
        Map<String, List<String>> groupAnagrams = new HashMap<>();
        
        
        for(String s: strs){
            
            char c[] = s.toCharArray();
            
            // sorting allows us to add every anagram to one key
            Arrays.sort(c);
            
            String newC = new String(c);
            
            if (!groupAnagrams.containsKey(newC)) {
                
                // this will create a key for every new anagram
                groupAnagrams.put(newC, new ArrayList<>());
            }
            // adds the orginal word to that key
            groupAnagrams.get(newC).add(s);
            
        }
        
      
        
        return new ArrayList<>(groupAnagrams.values());
        
        
    }
}

File: 49-group-anagrams/49-group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        
        dic = {}
        for string in strs:
            base = "".join(sorted(string)) # sorts the string 
            if base in dic:
                dic[base] += [string] # adds the string if it already exsists
            else:
                dic[base] = [string] # creates a new dict key
        return dic.values()
        
        
        
        
        
        
        
           
            